src/domains/queries.ml

```diff
@@ -329,3 +311,29 @@ struct
 
   let hash x = 31 * order x + hash_arg x
 end
+
+
+let eval_int_binop (module Bool: Lattice.S with type t = bool) binop (ask: ask) e1 e2: Bool.t =
+  let t1 = Cilfacade.typeOf e1 in
+  let t2 = Cilfacade.typeOf e2 in
+  let (_, e) = Cilfacade.doBinOp binop e1 t1 e2 t2 in
+  let i = ask.f (EvalInt e) in
+  if ID.is_bot i || ID.is_bot_ikind i then
+    Bool.top () (* base returns bot for non-int results, consider unknown *)
```

Comment:
Yes. Base itself uses the but result of an `EvalInt` query as indication that the expression actually doesn't evaluate to an integer type. It can then fall back to all the other cases. If it returned top instead, then it would never be able to distinguish, e.g. thread ID sets from actually unknown integers.

One reason for needing such inconsistency is, e.g. for these thread ID sets, because variables of `pthread_t` type are actually just integer values (on Linux), but we want base to evaluate them to `Thread` instead.